File: Java/pattern/FlyWeight/src/com/pattern/FlyWeightFactory.java
package com.pattern;

import java.nio.file.Files;
import java.util.HashMap;

public class FlyWeightFactory {
	private HashMap<Character, FlyWeight> flies;
	
	public FlyWeightFactory () {
		flies = new HashMap<>(10);
	}
	
	public FlyWeight factory(Character state) {
		if (flies.containsKey(state))
			return flies.get(state);
		else {
			FlyWeight flyWeight = new ConcreteFlyWeight(state);
			flies.put(state, flyWeight);
			return flyWeight;
		}
	}
}

File: cplusplus/design_pattern/FlyWeightPro/FlyWeightFactory.cpp
#include "stdafx.h"
#include "FlyWeightFactory.h"
#include "ConcreteFlyWeight.h"


FlyWeightFactory::FlyWeightFactory()
{

}


FlyWeightFactory::~FlyWeightFactory()
{
}

std::shared_ptr<FlyWeight> FlyWeightFactory::factory(std::string str)
{
    auto it = flies.find(str);
    if (it != flies.end())
        return flies[str];
    else {
        std::shared_ptr<FlyWeight> tempObj = std::make_shared<ConcreteFlyWeight>(str);
        flies.insert(std::make_pair(str, tempObj));
        return tempObj;
    }
}
